benches/dapol.rs

```diff
@@ -1,175 +1,385 @@
 
-fn verify_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("verify");
-    group.sample_size(10);
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    for h in TREE_HEIGHTS.into_iter() {
+        for t in thread_counts.iter() {
+            if *t > MaxThreadCount::default().get_value() {
+                continue;
+            }
+
+            for u in NUM_USERS.into_iter() {
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
 
-    // this benchmark depends on the tree height and not the number of leaves,
-    // so we just pick the smallest number of leaves
-    let num_leaves = NUM_LEAVES[0];
-    for &tree_height in TREE_HEIGHTS.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        let mut rng = thread_rng();
-        let item_range = Uniform::new(0usize, num_leaves);
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofSplitting>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("splitting", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
+                group.bench_function(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", &tup.0, &tup.1, &tup.2),
+                    ),
+                    |bench| {
+                        bench.iter(|| {
+                            ndm_smt = setup::build_ndm_smt(tup.clone()).ok();
```

Comment:
This will just swallow any legit errors that happen during tree build. Not that we expect which any errors, but if there is an error for some reason we wouldn't know. Wrt errors you should take this approach:
- if there is an error in the API then fail the whole bench suite
- if the input tuple is not valid then continue the execution